plugins/modules/helm_pull.py

```diff
@@ -0,0 +1,335 @@
+            aliases=["ssl_ca_cert"],
+            fallback=(env_fallback, ["K8S_AUTH_SSL_CA_CERT"]),
+        ),
+        validate_certs=dict(
+            type="bool",
+            default=True,
+            aliases=["verify_ssl"],
+            fallback=(env_fallback, ["K8S_AUTH_VERIFY_SSL"]),
+        ),
+        api_key=dict(
+            type="str", no_log=True, fallback=(env_fallback, ["K8S_AUTH_API_KEY"])
+        ),
+    )
+    module = AnsibleModule(
+        argument_spec=argspec,
+        supports_check_mode=True,
+        required_by=dict(
+            repo_username=("repo_password"),
+            repo_password=("repo_username"),
+        ),
+        mutually_exclusive=[("chart_version", "chart_devel")],
+    )
+
+    bin_path = module.params.get("binary_path")
+    if bin_path is not None:
+        helm_cmd_common = bin_path
+    else:
+        helm_cmd_common = "helm"
+
+    helm_cmd_common = module.get_bin_path(helm_cmd_common, required=True)
+
+    helm_version = get_helm_version(module, helm_cmd_common)
+    if LooseVersion(helm_version) < LooseVersion("3.0.0"):
+        module.fail_json(
+            msg="This module requires helm >= 3.0.0, current version is {0}".format(
+                helm_version
+            )
+        )
+
+    helm_pull_opt_versionning = dict(
+        skip_tls_certs_check="3.3.0",
+        chart_ca_cert="3.1.0",
+        chart_ssl_cert_file="3.1.0",
+        chart_ssl_key_file="3.1.0",
+    )
+
+    def test_version_requirement(opt):
+        req_version = helm_pull_opt_versionning.get(opt)
+        if req_version and LooseVersion(helm_version) < LooseVersion(req_version):
+            module.fail_json(
+                msg="Parameter {0} requires helm >= {1}, current version is {2}".format(
+                    opt, req_version, helm_version
+                )
+            )
+
+    # Set `helm pull` arguments requiring values
+    helm_pull_opts = []
+
+    helm_value_args = dict(
+        chart_version="version",
+        verify_chart_keyring="keyring",
+        repo_url="repo",
+        repo_username="username",
+        repo_password="password",
+        destination="destination",
+        chart_ca_cert="ca-file",
+        chart_ssl_cert_file="cert-file",
+        chart_ssl_key_file="key-file",
+    )
+
+    for opt, cmdkey in helm_value_args.items():
+        if module.params.get(opt):
+            test_version_requirement(opt)
+            helm_pull_opts.append("--{0} {1}".format(cmdkey, module.params.get(opt)))
+
+    # Set `helm pull` arguments flags
+    helm_flag_args = dict(
+        verify_chart=dict(key="verify"),
+        provenance=dict(key="prov"),
+        pass_credentials=dict(key="pass-credentials"),
+        skip_tls_certs_check=dict(key="insecure-skip-tls-verify"),
+        chart_devel=dict(key="devel"),
+        untar_chart=dict(key="untar"),
+    )
+
+    for k, v in helm_flag_args.items():
+        if module.params.get(k) and module.params.get(k) != v.get("default", False):
```

Comment:
This is still broken. If the default value were True and the user passed in True, or just left it unset, it's going to not add the flag. Can we just remove any comparison to a default value at all? I don't think there's any need for it, and it is adding unnecessary complexity. These are boolean values that control whether or not a flag is added to the command line. All we care about is whether the value we get in module.params is True or False. If it's True we set the flag, if it's False we don't.

Reply:
updated